test/jdk/java/nio/channels/Channels/TransferToBase.java

```diff
@@ -0,0 +1,252 @@
+
+        // tests input stream with a length between 1k and 4k
+        checkTransferredContents(inputStreamProvider, outputStreamProvider, createRandomBytes(1024, 4096));
+
+        // tests input stream with several data chunks, as 16k is more than a
+        // single chunk can hold
+        checkTransferredContents(inputStreamProvider, outputStreamProvider, createRandomBytes(16384, 16384));
+
+        // tests randomly chosen starting positions within source and
+        // target stream
+        for (int i = 0; i < ITERATIONS; i++) {
+            byte[] inBytes = createRandomBytes(MIN_SIZE, MAX_SIZE_INCR);
+            int posIn = RND.nextInt(inBytes.length);
+            int posOut = RND.nextInt(MIN_SIZE);
+            checkTransferredContents(inputStreamProvider, outputStreamProvider, inBytes, posIn, posOut);
+        }
+
+        // tests reading beyond source EOF (must not transfer any bytes)
+        checkTransferredContents(inputStreamProvider, outputStreamProvider, createRandomBytes(4096, 0), 4096, 0);
+
+        // tests writing beyond target EOF (must extend output stream)
+        checkTransferredContents(inputStreamProvider, outputStreamProvider, createRandomBytes(4096, 0), 0, 4096);
+    }
+
+    /*
+     * Special test for stream-to-file / file-to-stream transfer of more than 2 GB.
+     */
+    static void testMoreThanTwoGB(String direction, BiFunction<Path, Path, InputStream> inputStreamProvider,
+            BiFunction<Path, Path, OutputStream> outputStreamProvider,
+            ToLongBiFunction<InputStream, OutputStream> transfer) throws IOException {
+        // prepare two temporary files to be compared at the end of the test
+        // set the source file name
+        String sourceName = String.format("test3GBSource_transfer%s%s.tmp", direction,
+            RND.nextInt(Integer.MAX_VALUE));
+        Path sourceFile = CWD.resolve(sourceName);
+
+        try {
+            // set the target file name
+            String targetName = String.format("test3GBTarget_transfer%s%s.tmp", direction,
+                RND.nextInt(Integer.MAX_VALUE));
+            Path targetFile = CWD.resolve(targetName);
+
+            try {
+                // calculate initial position to be just short of 2GB
+                final long initPos = 2047*BYTES_PER_WRITE;
+
+                // create the source file with a hint to be sparse
+                try (FileChannel fc = FileChannel.open(sourceFile, CREATE_NEW, SPARSE, WRITE, APPEND)) {
+                    // set initial position to avoid writing nearly 2GB
+                    fc.position(initPos);
+
+                    // Add random bytes to the remainder of the file
+                    int nw = (int)(NUM_WRITES - initPos/BYTES_PER_WRITE);
+                    for (int i = 0; i < nw; i++) {
+                        byte[] rndBytes = createRandomBytes(BYTES_PER_WRITE, 0);
+                        ByteBuffer src = ByteBuffer.wrap(rndBytes);
+                        fc.write(src);
```

Comment:
Should this code perhaps allow for a short write, i.e., `FileChannel::write` returning a value less than `rndBytes.length`? For example, something like this:
```
                    long pos = initPos;
                    while (pos < BYTES_WRITTEN) {
                        int len = Math.toIntExact(BYTES_WRITTEN - pos);
                        if (len > BYTES_PER_WRITE)
                            len = BYTES_PER_WRITE;
                        byte[] rndBytes = createRandomBytes(len, 0);
                        ByteBuffer src = ByteBuffer.wrap(rndBytes);
                        pos += fc.write(src);
                    }
```

Reply:
I have incorporated your proposed code change.